src/frontend/src/optimizer/property/func_dep.rs

```diff
@@ -0,0 +1,338 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::fmt;
+
+use fixedbitset::FixedBitSet;
+use itertools::Itertools;
+
+use crate::utils::ColIndexMapping;
+
+/// [`FunctionalDependency`] represent a dependency of from --> to.
+#[derive(Debug, PartialEq, Clone, Default)]
+pub struct FunctionalDependency {
+    pub from: FixedBitSet,
+    pub to: FixedBitSet,
+}
+
+impl fmt::Display for FunctionalDependency {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let from = self.from.ones().collect_vec();
+        let to = self.to.ones().collect_vec();
+        f.write_fmt(format_args!("{:?} --> {:?}", from, to))
+    }
+}
+
+impl FunctionalDependency {
+    /// Create a [`FunctionalDependency`] with bitset.
+    /// This indicate a from --> to dependency.
+    pub fn new(from: FixedBitSet, to: FixedBitSet) -> Self {
+        assert_eq!(
+            from.len(),
+            to.len(),
+            "from and to should have the same length"
+        );
+        FunctionalDependency { from, to }
+    }
+
+    /// Create a [`FunctionalDependency`] with column indices.
+    pub fn with_indices(column_cnt: usize, from: &[usize], to: &[usize]) -> Self {
+        let from = {
+            let mut tmp = FixedBitSet::with_capacity(column_cnt);
+            for &i in from {
+                tmp.set(i, true);
+            }
+            tmp
+        };
+        let to = {
+            let mut tmp = FixedBitSet::with_capacity(column_cnt);
+            for &i in to {
+                tmp.set(i, true);
+            }
+            tmp
+        };
+        FunctionalDependency { from, to }
+    }
+
+    /// Create a [`FunctionalDependency`] for a key column. This column can determine all other
+    /// columns.
+    pub fn with_key_column(column_cnt: usize, key_column_id: usize) -> Self {
+        let mut from = FixedBitSet::with_capacity(column_cnt);
+        from.set(key_column_id, true);
+        let mut to = from.clone();
+        to.toggle_range(0..to.len());
+        FunctionalDependency { from, to }
+    }
+
+    /// Create a [`FunctionalDependency`] for a constant column. This column can be determined by
+    /// any column.
+    pub fn with_constant_column(column_cnt: usize, constant_column_id: usize) -> Self {
+        let mut to = FixedBitSet::with_capacity(column_cnt);
+        to.set(constant_column_id, true);
+        FunctionalDependency {
+            from: FixedBitSet::with_capacity(column_cnt),
+            to,
+        }
+    }
+
+    pub fn into_inner(self) -> (FixedBitSet, FixedBitSet) {
+        (self.from, self.to)
+    }
+}
+
+/// [`FunctionalDependencySet`] contains the functional dependencies.
+///
+/// It is used in optimizer to track the dependencies between columns.
+#[derive(Debug, PartialEq, Clone, Default)]
+pub struct FunctionalDependencySet {
+    strict: Vec<FunctionalDependency>,
```

Comment:
why this called `strict`? is there some comments?

Reply:
`strict` means strict dependency. We may also add lax dependency in the future. Lax dependency can be useful in outer joins.

Reply:
ok, so I think we need to add more comments here to explain that.